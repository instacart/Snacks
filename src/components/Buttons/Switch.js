import React from 'react'
import PropTypes from 'prop-types'
import RadioCheckboxBase from '../../base/RadioCheckboxBase'
import colors from '../../styles/colors'
import ZeroSvg from '../../assets/zero.svg'
import OneSvg from '../../assets/one.svg'

const SWITCH_WIDTH = 36

const STYLES = {
  background: {
    borderRadius: '11px',
    position: 'relative',
    transition: 'background-color 200ms ease-in-out',
  },
  toggle: {
    default: {
      height: '18px',
      width: '18px',
      backgroundColor: colors.WHITE,
      borderRadius: '50%',
      display: 'block',
      position: 'absolute',
      top: '2px',
      left: '2px',
      transition: 'left 200ms ease-in-out',
    },
    selected: {
      left: '16px',
    },
  },
  zero: {
    position: 'absolute',
    right: '5px',
    top: '7px',
  },
  one: {
    position: 'absolute',
    left: '8px',
    top: '7px',
  },
}

const renderInputButton = (isSelected, style, displayOnOffLabel) => {
  return (
    <div style={[style, { backgroundColor: style.fill }, STYLES.background]}>
      {displayOnOffLabel && isSelected && <OneSvg style={STYLES.one} />}
      {displayOnOffLabel && !isSelected && <ZeroSvg style={STYLES.zero} />}
      <div style={[STYLES.toggle.default, isSelected && STYLES.toggle.selected]} />
    </div>
  )
}

const Switch = props => {
  return (
    <RadioCheckboxBase
      btnType="checkbox"
      width={SWITCH_WIDTH}
      renderInputButton={(isSelected, style) =>
        renderInputButton(isSelected, style, props.displayOnOffLabel)
      }
      {...props}
    />
  )
}

Switch.propTypes = {
  aria: PropTypes.shape({
    label: PropTypes.string,
  }),
  children: PropTypes.node,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  isDisabled: PropTypes.bool,
  isSelected: PropTypes.bool,
  name: PropTypes.string,
  onChange: PropTypes.func,
  style: PropTypes.shape({
    button: PropTypes.object,
    label: PropTypes.object,
    wrapEl: PropTypes.object,
  }),
  value: PropTypes.string,
  wrapEl: PropTypes.string,
  displayOnOffLabel: PropTypes.bool,
}

export default Switch
