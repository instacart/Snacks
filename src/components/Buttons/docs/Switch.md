This component outputs a switch with an optional label.

You can define whether the button starts in the pressed on unpressed state through the `isSelected` prop. In addition, you can change the selected/unselected state at any moment after the mount state by redifining the prop `isSelected`.

### Label
Associating a label to every switch button is very important for accessibility purposes. There are two ways to do it. The easiest is to enclose the text with the component: `<Switch>My label</Switch>`. The other option is to associate a label yourself:

```jsx static
<div style={{display: 'flex', flexDirection: 'column'}}>
  <div style={{display: 'flex', marginBottom: 20}}>
    <Switch id="switch1" isSelected />
    <label htmlFor="switch1" style={{marginLeft: 10, lineHeight: '22px'}}>Label associated manually</label>
  </div>
  <Switch id="switch2">Label auto-generated</Switch>
</div>
```
**Important!** If you attach a label to a switch manually, make sure you set the line-height and margin-left properties as per the example above.

### On/Off Label
You can display an on off label to assist vision impaired users in determining whether the switch is on or off

```jsx static
<div style={{display: 'flex', flexDirection: 'column'}}>
  <Switch id="switch2" displayOnOffLabel>Switch with on/off label</Switch>
</div>
```

### Styling
You can pass an object to the `style` prop for styling the label, radio button and wrap element using Radium's structure:
```jsx static
<Switch
  id="switch3"
  wrapEl="span"
  style={{
    wrapEl: {padding: 10, border: '2px dotted lightblue'},
    button: {marginRight: 40},
    label: {color: 'salmon'},
  }}
>
  My Switch Button Label
</Switch>
```

### Callback
You can pass a callback through the `onChange` prop:

```jsx static
function onChange(event, props) {
  alert(`This button is ${props.isSelected ? '' : 'not '}selected`)
}

<Switch id="switch4" onChange={onChange}>Click me!</Switch>
```

### Disabling
You can disable a switch by declaring the `isDisabled` prop:

```jsx static
function onChange(event, props) {
  alert(`This function will not be called`)
}

<Switch id="switch5" onChange={onChange} isDisabled>I am disabled</Switch>
```
**Important!** If you attach a label to a switch manually, make sure the colors for the disabled state follow the styleguide.
