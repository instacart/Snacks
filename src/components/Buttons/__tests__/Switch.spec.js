import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import sinon from 'sinon'
import Switch from '../Switch'

describe('Switch', () => {
  it('renders the correct button type', () => {
    const tree = renderer.create(<Switch id={1} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders the correct selected state', () => {
    const testCases = [
      { id: 1, isSelected: false },
      { id: 2, isSelected: true },
      { id: 2, isSelected: true, isDisabled: true },
    ]

    testCases.forEach(props => {
      const tree = renderer.create(<Switch {...props} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('disables the input when it is disabled', () => {
    const wrapper = mount(
      <Switch id="A" value="test" isSelected>
        Testing disabled
      </Switch>
    )
    const htmlBtn = wrapper.find('input').instance()

    expect(htmlBtn.checked).toBe(true)
    expect(htmlBtn.disabled).toBe(false)

    wrapper.setProps({ isDisabled: true })

    expect(htmlBtn.checked).toBe(true)
    expect(htmlBtn.disabled).toBe(true)
  })

  it('generates a label if a child prop of text is supplied', () => {
    const tree = renderer.create(<Switch id={1}>Test label</Switch>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('incorporates user style if passed', () => {
    const testCases = [
      {
        id: 1,
        style: {
          wrapEl: { background: 'pink' },
          button: { border: '2px solid purple' },
          label: { color: 'green' },
        },
      },
      {
        id: 2,
        style: undefined,
      },
    ]

    testCases.forEach(props => {
      const tree = renderer.create(<Switch {...props}>Test label</Switch>).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('calls the user-supplied onChange function', () => {
    const onChange = sinon.spy()
    const wrapper = mount(<Switch id={1} onChange={onChange} />)

    wrapper.find('input').simulate('change')
    expect(onChange.calledOnce).toBe(true)
  })
})
