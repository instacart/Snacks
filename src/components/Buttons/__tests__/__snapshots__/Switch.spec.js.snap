// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch generates a label if a child prop of text is supplied 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#757575",
          "borderRadius": "11px",
          "fill": "#757575",
          "height": 22,
          "position": "relative",
          "transition": "background-color 200ms ease-in-out",
          "width": 36,
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": "50%",
            "display": "block",
            "height": "18px",
            "left": "2px",
            "position": "absolute",
            "top": "2px",
            "transition": "left 200ms ease-in-out",
            "width": "18px",
          }
        }
      />
    </div>
    <input
      checked={false}
      data-radium={true}
      id={1}
      onChange={[Function]}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "backgroundImage": "none",
          "cursor": "pointer",
          "height": 22,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 36,
        }
      }
      type="checkbox"
    />
  </div>
  <label
    data-radium={true}
    htmlFor={1}
    style={
      Object {
        "lineHeight": "22px",
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Switch incorporates user style if passed 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "background": "pink",
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "border": "2px solid purple",
        "position": "relative",
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#757575",
          "borderRadius": "11px",
          "fill": "#757575",
          "height": 22,
          "position": "relative",
          "transition": "background-color 200ms ease-in-out",
          "width": 36,
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": "50%",
            "display": "block",
            "height": "18px",
            "left": "2px",
            "position": "absolute",
            "top": "2px",
            "transition": "left 200ms ease-in-out",
            "width": "18px",
          }
        }
      />
    </div>
    <input
      checked={false}
      data-radium={true}
      id={1}
      onChange={[Function]}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "backgroundImage": "none",
          "cursor": "pointer",
          "height": 22,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 36,
        }
      }
      type="checkbox"
    />
  </div>
  <label
    data-radium={true}
    htmlFor={1}
    style={
      Object {
        "color": "green",
        "lineHeight": "22px",
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Switch incorporates user style if passed 2`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#757575",
          "borderRadius": "11px",
          "fill": "#757575",
          "height": 22,
          "position": "relative",
          "transition": "background-color 200ms ease-in-out",
          "width": 36,
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": "50%",
            "display": "block",
            "height": "18px",
            "left": "2px",
            "position": "absolute",
            "top": "2px",
            "transition": "left 200ms ease-in-out",
            "width": "18px",
          }
        }
      />
    </div>
    <input
      checked={false}
      data-radium={true}
      id={2}
      onChange={[Function]}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "backgroundImage": "none",
          "cursor": "pointer",
          "height": 22,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 36,
        }
      }
      type="checkbox"
    />
  </div>
  <label
    data-radium={true}
    htmlFor={2}
    style={
      Object {
        "lineHeight": "22px",
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Switch renders the correct button type 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "backgroundColor": "#757575",
        "borderRadius": "11px",
        "fill": "#757575",
        "height": 22,
        "position": "relative",
        "transition": "background-color 200ms ease-in-out",
        "width": 36,
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": "50%",
          "display": "block",
          "height": "18px",
          "left": "2px",
          "position": "absolute",
          "top": "2px",
          "transition": "left 200ms ease-in-out",
          "width": "18px",
        }
      }
    />
  </div>
  <input
    checked={false}
    data-radium={true}
    id={1}
    onChange={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "cursor": "pointer",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 36,
      }
    }
    type="checkbox"
  />
</div>
`;

exports[`Switch renders the correct selected state 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "backgroundColor": "#757575",
        "borderRadius": "11px",
        "fill": "#757575",
        "height": 22,
        "position": "relative",
        "transition": "background-color 200ms ease-in-out",
        "width": 36,
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": "50%",
          "display": "block",
          "height": "18px",
          "left": "2px",
          "position": "absolute",
          "top": "2px",
          "transition": "left 200ms ease-in-out",
          "width": "18px",
        }
      }
    />
  </div>
  <input
    checked={false}
    data-radium={true}
    id={1}
    onChange={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "cursor": "pointer",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 36,
      }
    }
    type="checkbox"
  />
</div>
`;

exports[`Switch renders the correct selected state 2`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "backgroundColor": "#43B02A",
        "borderRadius": "11px",
        "fill": "#43B02A",
        "height": 22,
        "position": "relative",
        "transition": "background-color 200ms ease-in-out",
        "width": 36,
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": "50%",
          "display": "block",
          "height": "18px",
          "left": "16px",
          "position": "absolute",
          "top": "2px",
          "transition": "left 200ms ease-in-out",
          "width": "18px",
        }
      }
    />
  </div>
  <input
    checked={true}
    data-radium={true}
    id={2}
    onChange={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "cursor": "pointer",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 36,
      }
    }
    type="checkbox"
  />
</div>
`;

exports[`Switch renders the correct selected state 3`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "backgroundColor": "#BDBDBD",
        "borderRadius": "11px",
        "fill": "#BDBDBD",
        "height": 22,
        "position": "relative",
        "transition": "background-color 200ms ease-in-out",
        "width": 36,
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": "50%",
          "display": "block",
          "height": "18px",
          "left": "16px",
          "position": "absolute",
          "top": "2px",
          "transition": "left 200ms ease-in-out",
          "width": "18px",
        }
      }
    />
  </div>
  <input
    checked={true}
    data-radium={true}
    disabled={true}
    id={2}
    onChange={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "cursor": "pointer",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 36,
      }
    }
    type="checkbox"
  />
</div>
`;
